sshnode.go

```diff
@@ -103,22 +120,45 @@ func newCmdStrWithSource(cmd string, env []string) string {
 	return command
 }
 
-func (n *SSHNode) getClientAndSession() (*ssh.Client, *ssh.Session, error) {
-	var client *ssh.Client
-	var s *ssh.Session
-	var err error
-
-	// Retry few times if ssh connection fails
+func (n *SSHNode) connect() error {
 	for i := 0; i < MaxSSHRetries; i++ {
-		client, err = n.dial()
+		client, err := n.dial()
 		if err != nil {
 			time.Sleep(SSHRetryDelay)
 			continue
 		}
+		n.client = client
+	}
+	if n.client == nil {
+		return fmt.Errorf("failed to connect to host")
+	}
 
-		s, err = client.NewSession()
+	return nil
+}
+
+func (n *SSHNode) getSession() (*ssh.Session, error) {
+	var s *ssh.Session
+	var err error
+
+	n.clientMutex.RLock()
+	if n.client == nil {
+		n.clientMutex.RUnlock()
+		n.clientMutex.Lock()
+		n.connect()
+		n.clientMutex.Unlock()
+	} else {
+		n.clientMutex.RUnlock()
+	}
+
+	for i := 0; i < MaxSSHRetries; i++ {
+		n.clientMutex.RLock()
+		s, err = n.client.NewSession()
+		n.clientMutex.RUnlock()
 		if err != nil {
-			client.Close()
+			n.clientMutex.Lock()
+			n.client.Close()
```

Comment:
Are we sure that every possible error returned from NewSession() requires a Close() and connect() ?